Add muli-threading support description to the manual (#2823)
JuliaData/DataFrames.jl: docs/src/lib/functions.md

This is a list of operations that currently support multi-threading:
- `DataFrame` constructor with `copycols=true`; also recursively all functions
  that call this constructor, e.g. `copy`.
- `getindex` when multiple columns are selected.
- `groupby` (both when hashing is required and when fast path using `DataAPI.refpool`
  is used).
- `join*` functions for composing output data frame (but currently not for finding
  matching rows in joined data frames).
- in `combine`, `select[!]``, and `transform[!]` on `GroupedDataFrame` when performing
  multiple transformations (each transformation is spawned in a separate task)
- in `combine` when a transformation produces one row per group and the passed tranformation
  is a custom function (i.e. when reduction based aggregation is not performed).

In general at least Julia 1.4 is required to ensure that multi-threading is used
and the Julia process must be started with more than one thread. Some operations
turn on multi-threading only if enough rows in the processed data frame is present
(the exact threshold when multi-threading is enabled is considered to be undefined
and might change in the future).

> 

Author: ```suggestion

Except for the list above, where multi-threading is used automatically, all functions provided by DataFrames.jl that update a data frame are not thread safe.
```

This is a list of operations that currently support multi-threading:
- `DataFrame` constructor with `copycols=true`; also recursively all functions
  that call this constructor, e.g. `copy`.
- `getindex` when multiple columns are selected.
- `groupby` (both when hashing is required and when fast path using `DataAPI.refpool`
  is used).
- `join*` functions for composing output data frame (but currently not for finding
  matching rows in joined data frames).
- in `combine`, `select[!]``, and `transform[!]` on `GroupedDataFrame` when performing
  multiple transformations (each transformation is spawned in a separate task)
- in `combine` when a transformation produces one row per group and the passed tranformation
  is a custom function (i.e. when reduction based aggregation is not performed).

In general at least Julia 1.4 is required to ensure that multi-threading is used
and the Julia process must be started with more than one thread. Some operations
turn on multi-threading only if enough rows in the processed data frame is present
(the exact threshold when multi-threading is enabled is considered to be undefined
and might change in the future).

> Except for the list above, where multi-threading is used automatically, all functions provided by DataFrames.jl that update a data frame are not thread safe.

Author: ```suggestion
Except for the list above, where multi-threading is used automatically,
all functions provided by DataFrames.jl that update a data frame are not thread safe.

```

# Functions

## Multi-threading support

Selected operations in DataFrames.jl support muliti-threading.
It is task-based and implemented using the `@spawn` macro form Julia Base.

This is a list of operations that currently support multi-threading:
- `DataFrame` constructor with `copycols=true`; also recursively all functions
  that call this constructor, e.g. `copy`.
- `getindex` when multiple columns are selected.
- `groupby` (both when hashing is required and when fast path using `DataAPI.refpool`
  is used).
- `join*` functions for composing output data frame (but currently not for finding
  matching rows in joined data frames).
- in `combine`, `select[!]``, and `transform[!]` on `GroupedDataFrame` when performing
  multiple transformations (each transformation is spawned in a separate task)

> - in `combine` when a transformation produces one row per group and the passed tranformation

Reviewer: Any comment on this?

> - in `combine` when a transformation produces one row per group and the passed tranformation

Author: Ah - I missed the question. I was thinking if we should include this. This can happen only if `select` or `transform` are passed a `GroupedDataFrame` that has exactly one row per group and when aggregation also produces one row per group. This is an unlikely scenario (in this scenario it does make sense to do `groupby` and then `select`/`transform` but rather one should just run `select`/`transform` on a data frame with `ByRow`). Do you think we should add `select` and `transform` to the list (as technically they could, but the condition when they would is unlikely).

In essence since this is "one row per group" case it means that we are `combine`ing rows unless each group has exactly one row to start with 😄.

> - in `combine` when a transformation produces one row per group and the passed tranformation

Author: @nalimilan - what do you think given this comment?

> - in `combine` when a transformation produces one row per group and the passed tranformation

Reviewer: Sorry, I had missed your reply. Isn't multithreading also used with `select` and `transform` when it returns a single row for each group (disregarding the number of rows in the group)? We repeat the returned row to match the number of rows in the input, but the step before that is multithreaded IIRC, right?

> - in `combine` when a transformation produces one row per group and the passed tranformation

Author: You are right (as usual - I am glad to have you :smile:). I have fixed this.

>   is a custom function (i.e. when reduction based aggregation is not performed).

Reviewer: ```suggestion
  is a custom function (i.e. not for standard reductions, which use
  optimized single-threaded methods).
```

In general at least Julia 1.4 is required to ensure that multi-threading is used
and the Julia process must be started with more than one thread. Some operations

> turn on multi-threading only if enough rows in the processed data frame is present

Reviewer: ```suggestion
turn on multi-threading only if enough rows in the processed data frame are present
```

(the exact threshold when multi-threading is enabled is considered to be undefined
and might change in the future).

Except for the list above, where multi-threading is used automatically,

> all functions provided by DataFrames.jl that update a data frame are not thread safe.

Reviewer: ```suggestion
all functions provided by DataFrames.jl that update a data frame are not thread safe.
This means that while they can be called from any thread, the caller is responsible
for ensuring that a given `DataFrame` object is never modified by one thread while
others are using it (either for reading or writing). Using the same `DataFrame`
at the same time from different threads is safe as long as it is not modified.
```

# Functions

## Multi-threading support

Selected operations in DataFrames.jl automatically use multiple threads when available.
It is task-based and implemented using the `@spawn` macro from Julia Base.

This is a list of operations that currently make use of multi-threading:
- `DataFrame` constructor with `copycols=true`; also recursively all functions
  that call this constructor, e.g. `copy`.
- `getindex` when multiple columns are selected.
- `groupby` (both when hashing is required and when fast path using `DataAPI.refpool`
  is used).
- `join*` functions for composing output data frame (but currently not for finding
  matching rows in joined data frames).

> - `combine`, `select[!]``, and `transform[!]` on `GroupedDataFrame` when performing

Author: ```suggestion
- `combine`, `select[!]`, and `transform[!]` on `GroupedDataFrame` when performing
```

> - `combine`, `select[!]`, and `transform[!]` on `GroupedDataFrame` when performing
>   multiple transformations (each transformation is spawned in a separate task)
> - `combine` when a transformation produces one row per group and the passed transformation
>   is a custom function (i.e. not for standard reductions, which use
>   optimized single-threaded methods).

Author: ```suggestion
- `combine`, `select[!]`, and `transform[!]` on `GroupedDataFrame` when either of the conditions below is met:
  * performing multiple transformations (each transformation is spawned in a separate task)
  * a transformation produces one row per group and the passed transformation
    is a custom function (i.e. not for standard reductions, which use
    optimized single-threaded methods).
```
